src/rvsdg/simplify_branches.rs

```diff
@@ -0,0 +1,623 @@
+/// use these as a heuristic to find blocks that are safe to merge into their
+/// predecessors in exchange for simpler control flow: RVSDG conversion overhead
+/// is largely contained in blocks only containing these instructions.
+fn is_admin_instr(inst: &Instruction) -> bool {
+    matches!(
+        inst,
+        Instruction::Constant { .. }
+            | Instruction::Value {
+                op: ValueOps::Id,
+                ..
+            }
+    )
+}
+
+fn constants_assigned(inst: &Instruction) -> Option<(Identifier, Literal)> {
+    if let Instruction::Constant {
+        dest,
+        value: value @ Literal::Bool(..),
+        ..
+    } = inst
+    {
+        Some((dest.into(), value.clone()))
+    } else {
+        None
+    }
+}
+
+fn to_lit(cv: &CondVal) -> Option<Literal> {
+    if cv.of == 2 {
+        Some(if cv.val == 0 {
+            Literal::Bool(false)
+        } else {
+            Literal::Bool(true)
+        })
+    } else {
+        // Not handling multi-way branches for now.
+        None
+    }
+}
+
+/// A basic semilattice describing the state of a value.
+#[derive(Clone, Default, Debug)]
+enum ValueInfo {
+    #[default]
+    Bot,
+    Known(Literal),
+    Top,
+}
+
+impl ValueInfo {
+    fn merge(&mut self, other: &ValueInfo) -> bool {
+        match (self, other) {
+            (ValueInfo::Bot, ValueInfo::Bot) => false,
+            (slf @ ValueInfo::Bot, x) => {
+                *slf = x.clone();
+                true
+            }
+            (ValueInfo::Top, _) => false,
+            (slf, ValueInfo::Top) => {
+                *slf = ValueInfo::Top;
+                true
+            }
+            (ValueInfo::Known(l), ValueInfo::Known(r)) if l == r => false,
+            (slf @ ValueInfo::Known(_), ValueInfo::Known(_)) => {
+                *slf = ValueInfo::Top;
+                true
+            }
+            (ValueInfo::Known(_), ValueInfo::Bot) => false,
+        }
+    }
+}
+
+/// Monotone transforms on ValueInfos.
+#[derive(Debug)]
+enum Transform {
+    Id,
+    Negate,
+    OverWrite(ValueInfo),
+}
+
+impl Transform {
+    fn apply(&self, val: &ValueInfo) -> ValueInfo {
+        match self {
+            Transform::Id => val.clone(),
+            Transform::Negate => match val {
+                ValueInfo::Bot => ValueInfo::Bot,
+                ValueInfo::Known(Literal::Bool(b)) => ValueInfo::Known(Literal::Bool(!b)),
+                ValueInfo::Known(..) => ValueInfo::Top,
+                ValueInfo::Top => ValueInfo::Top,
+            },
+            Transform::OverWrite(info) => info.clone(),
+        }
+    }
+}
+
+/// The state of the (boolean) values in a particular basic block.
+#[derive(Default, Debug)]
+struct ValueState {
+    /// The (pointwise) join of all of the values in incoming branches.
+    inherited: IndexMap<Identifier, ValueInfo>,
+    /// The transforms induced by any operations on variables in the block.
+    transforms: VecDeque<(
```

Comment:
comment: computed once, on initialization. Then virtual assignments are added to it during step 2